packages/core/core/src/Strapi.ts

```diff
@@ -269,20 +270,19 @@ class Strapi extends Container implements Core.Strapi {
       .add('entityValidator', entityValidator)
       .add('entityService', () => createEntityService({ strapi: this, db: this.db }))
       .add('documents', () => createDocumentService(this))
-      .add(
-        'db',
-        () =>
-          new Database(
-            _.merge(this.config.get('database'), {
-              logger,
-              settings: {
-                migrations: {
-                  dir: path.join(this.dirs.app.root, 'database/migrations'),
-                },
+      .add('db', () => {
+        const outDir = tsUtils.resolveOutDirSync(this.dirs.app.root) ?? this.dirs.app.root;
+        return new Database(
+          _.merge(this.config.get('database'), {
+            logger,
+            settings: {
+              migrations: {
+                dir: path.join(outDir, 'database/migrations'),
```

Comment:
```suggestion
        const tsDir = tsUtils.resolveOutDirSync(this.dirs.app.root);
        const projectDir = this.config.get('database.useTypescriptMigrations') === true && tsDir ? tsDir : this.dirs.app.root;
        return new Database(
          _.merge(this.config.get('database'), {
            logger,
            settings: {
              migrations: {
                dir: path.join(projectDir, 'database/migrations'),
```

Something like this might be the best option for now to avoid breaking anything (note: I didn't test it, just wrote it here quickly to show the idea)

As a follow-up, we will start generating new typescript projects with `database.useTypescriptMigrations:true` already set, and in Strapi 6 we can make it the default behaviour.

Let's get some opinions on if this is the best place to put it, and then we will need to document it.

cc @remidej @Marc-Roig @butcherZ

Reply:
Also tagging @alexandrebodin in case you have any comment on it

Database config seems the right place for it, my only concern is that we've never put an option other than `connection` there before afaik.

Reply:
@DevDengChao could you please have it only use the new behaviour when config `database.settings.useTypescriptMigrations` is true? I'll create a docs PR for it